### DISCUSSION TITLE
[Feature Request] Nested tags

### DISCUSSION THREAD

**Discussion by woodwm** [Future-Scholars/paperlib#354] [Ideas]
Obsidian supports [nested tags](https://help.obsidian.md/Plugins/Tags) using `#inbox/to-read`. Nested tags can replace folders to show hierarchy relationship.

**Comment by GeoffreyChen777** [comment]
This requires a huge change of the database and the UI component. I’m afraid that I can’t implement it in a short time.

**Comment by mrzzmrzz** [comment]
I think the feature is really necessary, support this idea!

**Comment by GeoffreyChen777** [comment]
# The following idea has been deprecated.


Hi @woodwm @Mrz-zz @fkcptlst @aur3l14no

Recently, I have a good idea to support nested tags/folders, but in an elegant way. I want to hear from you to see if it's really a good idea.

## Atomic Design
The basic design of Paperlib's tag system is to keep it simple as I want to use the simplest way to organize papers. Previously we didn't support nested tags/folders due to the **atomicity**. A tag/folder should be the smallest element in the system. For example, the tags of a paper related to 'semi-supervised image classification' should be `semi-supervised` and `classification`.

## SmartFilter
After reading some user feedback such as #237, I realised that combining atomic tags/folders to create a smartfilter is a good way to filter papers related to more than one topic such as the abovementioned one. The smartfilter has been introduced for a while and I personally love it.

## New Idea
Although smartfilter is a good workaround to create combined tags/folders, It still looks far away from nested tags/folders.
Recently, after discussing with @charlieJ107, I have a good idea for the tag/folder system:

 **Parse the rule of a smartfilter into a graph and show it in the UI.**

Suppose we have a smartfilter like this:
```
(tags.name == "semi-supervised") AND (tags.name == "classification")
```
It can help us to find the papers that have both two tags.

This smartfilter query can be transformed into a graph:

<img width="421" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/26cd2af7-3820-402f-84a9-b055d220c45f">

Apparently, even a complex query can also be transformed into a graph:
```
(tags.name == "semi-supervised") AND (tags.name == "classification" OR tags.name == "segmentation")
```
<img width="545" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/5f936725-8f0c-464c-91cb-f8528870ff91">

If we walk start from the `Root Node`, we can get a view tree according to the nodes' depth:
```
// Tree of the graph1
| - Root Node
     | - Statement: tags.name == "semi-supervised"
     | - Statement: tags.name == "classification"
```
```
// Tree of the graph2
| - Root Node
     | - Statement: tags.name == "semi-supervised"
     | - Logic Node: OR
         | - Statement: tags.name == "classification"
         | - Statement: tags.name == "segmentation"
```

We can show this tree after you click the corresponding smartfilter like this:
<img width="308" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/baca1fb0-f08d-4ea2-96f3-39ea84691fe4">

Click a node in this tree can filter the database by the corresponding filter statement.

By doing so, the nested tags/folders can be achieved without changing the database structure. And we still have the atomicity.

---

I appreciate all your responses. 👍

**Reply by GeoffreyChen777** [reply]
<img width="886" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/7ea256fb-8b90-472e-be02-c6591771e677">

**Reply by GeoffreyChen777** [reply]
I agree that these two kinds of trees are different.

Showing a syntax tree is a compromise for nested tags/folders as I really don't want to discard the atomicity (and redesign the database structure😵).

When a logic OR node's children consist of only tags/folders, it equals a parent set of them. If we can rename this logic node, that would be the best way. But I think it's hard to implement.

What about this solution:

For the example shown above, when we show the tree, the logic OR is rendered as segmentation / classification rather than Logic OR.

If the logic node is an AND, we render segmentation & classification

**Reply by GeoffreyChen777** [reply]
<img width="885" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/24a84da0-b1c3-43c5-a1d4-2f3d17fd2208">

<img width="892" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/1e80fdfa-21ab-4cf3-b447-6d84a5b1009f">

**Reply by GeoffreyChen777** [reply]
That's interesting. I want to discuss more about your design:

1. Suppose I have a Paper C, it is a multi-modal segmentation paper. So I should have tags like:
Paper C: `multi-modal/segmentation`.
When dealing with tags, is the `segmentation` under `multi-modal` identical to the one under `cv`?

2. In the database, we store something like:
  ```
  {
    name: cv
    ...
  }
  
  {
    name: cv/segmentation
    ...
  }
  ```
  
  or we store:
  ```
  {
    name: cv
    ...
  }
  
  {
    name: segmentation
    ...
  }
  ```
  i.e., the atomicity should be kept only in UI or in both UI and database?

I think, for folders, your solution is really good. Because two subfolders under two different parent folders should be treated as two identities even if they have exactly the same name.

But for tags, I think due to the atomicity, they should not be treated as two identities. The `segmentation` tag should be considered the same whether it is in hyper-tag A (e.g., `cv`) or hyper-tag B. (e.g., `multi-modal`). If we store the tags like this:
```
{
  name: cv/segmentation 
  ...
}
```

, when we want to find papers related to only `segmentation`, the database need to give me the results with tags like:

```
{
  name: segmentation    ✅
  ...
}

{
  name: cv/segmentation     ✅
  ...
}

{
  name: multi-modal/segmentation     ✅
  ...
}

{
  name: cv/classification     ❌
  ...
}
```
It means that in the database, we have no atomicity right?

**Reply by GeoffreyChen777** [reply]
<img width="894" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/95481528-52c7-46db-8b9e-89afa8b4465a">

**Comment by charlieJ107** [comment]
I agree that by creating a new table to store complex tag or folder relationships, user-defined smart filters can also be converted into such tag relationship entries. When querying, first generate query conditions based on tag or folder relationships, and then use these query conditions to query the database. This can preserve the atomicity of storage in the database.

In fact, this is equivalent to abandoning the original concept of folders. Folders are regarded as a special label. The so-called nested relationship between folders and labels becomes an AND logic between labels. I'm not sure whether we should embody this concept in the UI from a user experience perspective. For example, we can simply remove the concept of folders.

**Reply by GeoffreyChen777** [reply]
Hi, let's discuss it from three aspects: User Aspect (UI), Design Aspect, and Implementation Aspect.

## User Aspect

I think the UI design I discussed with @fkcptlst is good, i.e., we have two modes for the left sidebar panel: 
   - Flat Mode: just like what we have in the current version.
   - Hierarchy Mode: show a tree

For `folders`, previously I just used it for the link-to-folder feature of the quick-paste plugin (the window shown when we press `cmd+shift+I`. For user-friendly, I think at least we need to keep the `tag` concept as it is very intuitive.

I agree that we need to reconsider the `categorizer/label system` in Paperlib.

## Design Aspect

In this section, I will propose a new design for the `categorizer/label system` in Paperlib.

Basically, in the current Paperlib, we have three concepts related to `categorizers`:

- tag
- folder
- smartfilter

The first two are user-defined properties of a paper. The third one is a complex query sentence.

So if we decide to change the database structure, we can modify the `PaperEntity` model like this:
```typescript
interface PaperEntity {
  title: string;
  authors: string;
  ...
  userProps: {
    tags: string[]
    folders: string[]
  };
  ...
}
```
We can use `userProps.tags` and `userProps.folders` to represent the first two concepts. And we only store their names. About the colour, I will discuss it later. Here we also have the possibility to add more user-preferred properties in the future.

In the following, I ignore `userProps` for simplicity: `paper.tags` `paper.folders`.

In the sidebar panel, previously we have three sections: `Smart Filters`, `Tags` and `Folders`. As they are all real models in the database, we just show them directly.

However, each row in the sidebar panel can be regarded as a `query sentence`:

- tag: A query sentence for the corresponding tag, such as `tags.name == 'Tag1'`
- folder: A query sentence for the corresponding folder, such as `folder.name == 'Folder1'`
- smartfilter A complex query, such as `tags.name == 'Tag1' and folders.name == 'Folder1'`

As you mentioned that you think we can use a graph for the relationships between categorizers, what about we regard `query sentence` as nodes in a graph. The edge between two nodes indicates how we combine them. By doing so, we can achieve the nested tags/folders, and most importantly abstract tags, folders, and smartfilters into a single concept.

### `Graph`

It should be a directed graph. 

### `Node`

<img width="246" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/be1e4f96-e419-4fce-9647-245152be8cb1">

The `Node` structure consists of three main properties: `Name`, `Self Query`, and `In Edge Logic` (I just ignore the `id` for simplicity):

- `Name`: indicates the string we show in the UI.
- `Self Query`: is the `query sentence` of this node.
- `In Edge Logic`: indicates how to combine the queries of those nodes pointing to this node.

So in a graph. We have three kinds of nodes:
<img width="587" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/5597ae00-9060-4980-b343-2dc29ce6fac3">
They are just for explanation. In the implementation/database, we can use one class/model for them.
- **Logic Node**: contains no self-query (`true`). Just use it for node combining.
- **Query Node**: contains a `query sentence`. But it will never gather and combine the `query sentences` of the nodes pointing to this node, as its `In Edge Logic` is just a `true`
- **Mix Node**: mixture of the above two types.

### `Graph` for Nested Tags

<img width="721" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/1d68592f-d97b-4713-ae96-663dc27e35bc">

Suppose we have three tags as shown in the figure, in the graph, we should have three `Query Nodes`. In addition, we have a `ROOT Node` Tags. That is the one shown here:

<img width="310" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/4dd8fcb7-23ed-42fc-b853-bcac9b502770">

The nested structure is represented by the topology.

As the tags concept in Paperlib is atomic, we can easily transform them into a simple `query sentence` like `'xxx' in paper.tags`

If the user prefers the Flat Mode, we can just get the entire graph connected with the `Tags ROOT Node` and flat it to show.

If the user prefers the Hierarchy Mode, we can transform this graph into a tree and show it in the UI:

<img width="743" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/c980c385-3213-41da-8608-40ee3310a61e">

Here we can see that the `In Edge Logic` of all nodes in this graph is just `true`. This is due to the way we create the final query when users click a tag in the sidebar.

#### Generate the final query from a graph

Suppose now we click the `cv` tag (Query Node 1) in the sidebar. The results we want should be all papers related to `cv`. We don't care about the fine-grained research topics. Even a paper containing only a single `cv` tag should be included.

To generate the final query from a graph, in this case, we start from the `Query Node1`:
- We first get the `Self Query` of this node. Now we have `'cv' in paper.tags`.
- Next, we iteratively go to the upper-level node until the `ROOT Node`, get all queries in these nodes and combine them with `AND`. Now we have `'cv' in paper.tags AND true`
- Next we aggregate the lower-level node according to the `In Edge Logic`. If it is a string with some placeholders such as `$1 OR $2`, we collect corresponding queries. If it is just a `true`, we will not collect the queries of the bottom-level nodes. We finally combine the queries we collected with the previous one with `AND`, now we have: `'cv' in paper.tags AND true AND true`.
`'cv' in paper.tags AND true AND true` is the final query we should use to filter the database when we click the `cv` tag.

If we click the `segmentation` tag under the `cv` tag, we should have a final query like this:
```
'segmentation' in paper.tags AND 'cv' in paper.tags AND true AND true
```
It will give us the desired results: papers related to segmentation and cv.

For `AND true`, we can simply ignore it in implementation.

### `Graph` for Smart Filters

For smart filters, we can also represent them as a graph. I basically have two designs, I really don't know which one is better:

#### Design 1

<img width="1143" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/1ecd56c3-4274-490c-bce5-0af60795cf3e">

The ROOT node is used to get the graph related to smart filters. We will also show it as a section in the sidebar.

Suppose we have two kinds of smart filters: a nested one (`last two years`), and a single one (`cv paper by Author A`).

For the nested one, we create a tree as shown in the figure. The main difference with the tree for tags is that, some `In Edge Logic` is a rule string, not a `true`. It will help us to generate the correct final query.

For example, when we click `last two years - Author A` in the sidebar (Mix Node 2 in the graph), we want to get the papers written by Author A, and in the recent two years, and only related to `optimisation` and `segmentation`. **That is different from what we want in tags.** In tags, we don't care about the lower-level node by using a `true` as the `In Edge Logic`. But in a smart filter, I think we should care about the lower-level node. We may only want to use this smart filter to find papers about two topics written by Author A. Author A may have other papers in our library, but we don't care about them in this smart filter.

Thus, we need to modify the `In Edge Logic` to `$0 OR $1`. By performing the process I described above, we can have a final query: `paper.authors contains 'Author A' AND (paper.pubTime == 2024 OR paper.PubTime == 2023) AND ('segmentation' in paper.tags OR 'optimisation' in paper.tags)`.

If a user wants to create a single complex smart filter, we can create a node like`Query Node 1`.

#### Design 2

This design is similar to what we do for tags. It means that all `In Edge Logic` is just a single `true`. When we click `last two years of AB - Author A`, we get all papers written by Author A in the last two years. We don't care about the lower-level nodes here.

---

I really don't know which one is better. Maybe we should keep both of them and make it configurable?

### Folders

~~For folders, it's similar to tags, but with a different ROOT Node.~~

I just realised that folders are totally different.

```
| -- Folder A
    | -- abc
| -- Folder B
    | -- abc
```

Because these two `abc` under different parent folders are totally different in a file system. It means that the real name of `abc` is `Folder A/abc`, not simply `abc` as we have in a tag system.

<img width="1132" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/c2e4f975-a7d8-42e4-84a7-fbe060fa1dd7">

So we probably need to store `xxx/yyy` as the name of each folder.

### Colours

We have different colours for tags/folders/smartfilters. In the new design, we store the colour attribute in each node.

---

By doing so, we can use graphs to represent the concept we used in Paperlib and their relationship in a uniform way. And we still have the atomicity.

Users may create their own ROOT node rather than the basic `Tags` / `Folders` / `Smart Filters`.

## Implementation Aspect

1. I need to investigate the database to see how to store the relationship between objects.
2. Implement a tree view for the sidebar.
3. Create a function to generate the final query according to a `Graph` and the starting `Node`.
4. We need to design an easy-to-use UI for users to create `Node` and `Graph`. For users, they should not access/understand the concept of `Node` and `Graph`. They should just feel like they are creating hierarchical tags / folders / smartfilters. Most importantly, it should be a **top-to-bottom** creation.
5. How to count the number of papers related to a node is a problem.

**Reply by GeoffreyChen777** [reply]
@charlieJ107  I have implemented a demo for the proposed `graph-based-filtering-system`:

https://github.com/Future-Scholars/demo-graph-based-filter

It looks good.

But I found some limitations of database model design.

Realm doesn't support define model like this:

```typescript
interface Paper {
  title: string;
  ...
  userProps: {
    tag: string[]
  };
}
```

I have no choice but to define it like this:

```typescript
interface Paper {
  title: string;
  ...
  userProps: UserProp[];
}

interface UserProp {
  key: string;
  value: any[]
}
```
However, the query sentence for tags/folders looks very complex:
```
SUBQUERY(userProps, $item, $item.key == 'tags' AND 'optimisation' in $item.value).@count > 0
```

We need to find a better way for this.

The graph-based query generation looks good.

**Comment by GeoffreyChen777** [comment]
> For db implementation, I think storing `cv/segmentation` would suffice (it is not so pretty since it sacrifices the atomicity of data table, but it gets the job done).
> 
> Or you can create a new table to store the hierarchy alone. The advantage of this method is that it does not break the process logic of original tags. The tags table can maintain its atomicity.

@fkcptlst So now I understand your design:

- For design/concept aspect, tags should be atomic.
- For database:
  - we store hierarchical names like `cv/segmentation`. If the user also has a top-level tag `segmentation`, or a child-tag under a different hyper-tag such as `multi-modal/segmentation`, we store all of them in the database. Consequently, in this case, we have 3 objects in the database, but in the top-level design aspect, they indicate the same concept.
- For UI:
  - In the right details panel, main table view, tags should be shown in an atomic way like what we have now: `cv, segmentation`
  - In the left sidebar panel, it depends on the user preference. We should have two modes:
    - Flat Mode: just like what we have in the current version.
    - Hierarchy Mode: show a tree rendered according to the hierarchy name
- For querying:
  - In the Flat Mode, if a user wants to show all papers related to a tag such as `segmentation`, the filtered results should be those with tags `segmentation` or `cv/segmentation` or `multi-modal/segmentation`
  - In the Hierarchy Mode, we just show the results with the exact hierarchy tags.


Pls correct me if I'm wrong.

**Comment by woodwm** [comment]
Hi, the discussion is quite long. Let me explain what I think of the usage and design of nested tag.

I found a good reason of using nested tag in [Evernote](https://help.evernote.com/hc/en-us/articles/4412905761299-Create-nested-tags), 

> If you have a lot of tags in Evernote, you can organize them by creating nested tags. Nested tags are tags that are organized as sub-tags under a main tag. With nested tags, there is a main tag and one or more sub-tags. For example, a main tag "Travel" might have multiple sub-tags "Chicago," "New York," and "San Francisco."

This is a good and clear explaination of nested tag which is used to **organize tags**. Considering this, we are ready to think about `cv/segmentation` and `multi-modal/segmentation`. Does the `segmentation` in the two nested tags the same or not? My answer is no. The only common part is the name, but they refer to different concepts. The same applies to `misc` in `notes/misc`, `supplement/misc`. If we agree that they are different, their tag need to be different.

This can be done by define a UUID for each tag. Tag has the property of `name`. A new object `nested_tag` is a instance of `tag`, but has addtional property of `parient tag` and `child tag`. `nested_tag` is more general than `tag`, as normal tag can be considered as a nested tag with no `parient tag` and `child tag` which is a top-level tag. `cv/segmentation` is a nested tag with two nested tag objects `cv` and `segmentation`. `segmentation` in `cv` and `multi-modal` are different objects with different UUIDs but the same name.

The good point of this design is that when the user add a tag `segmentation` to a paper, we can provide suggestions to allow it to choose which `segmentation` the user refers to and select either `cv/segmentation` or `multi-modal/segmentation` and complete the other parts for the user.

In this concept, the Flat Mode does not make sense as the conflict in tag name. The nested tag becomes first cityzen. We can still show all papers that have `segmentation` tag by using smartfilter.

**Reply by GeoffreyChen777** [reply]
Hi, in your example, you think that `notes/misc` is different from `supplement/misc`. I think the reason is that you don't consider the **atomicity** of tags.

 In this case, you regard `note/misc` as a single tag: `note-misc`. However, it should be a joint hyper-tag consisting of two atomic tags --- `note` and `misc`.

In a tag system, we should have the **atomicity** which is very intuitive. Your design is basically a file/folder system, not a tag system.

Let me extend your travel example to explain why we need **atomicity** in a tag system, suppose we have some tags in **non-atomic style**, and their hierarchy is like this:
```
| -- Travel
    | -- New York
    | -- Chicago
```

Suppose we have some new photos/documents:

- A photo of Times Square: we should assign `Travel/New York` to it. When we want to retrieve this photo, we will first go to `Travel` level, then go to `New York` level, and finally we find it. right?

Now we also have some photos of delicious foods in NYC. And I am a gourmet. Then I may have a hierarchy like this:

```
| -- Travel
    | -- New York
    | -- Chicago
| -- Tasty
    | -- New York
    | -- Beijing
    | -- Tokyo
    | -- ...
```

- Suppose during the journey to NY, I took some photos of the foods of NY. Now I should assign `Tasty/New York` to these photos. When I want to find these photos, I need to go to `Tasty - New York`, right? When I click this sub tag, the results should only contain food photos, no travel photos right?

Now, What if I also have a top-level `New York`:

```
| -- Travel
    | -- New York
    | -- Chicago
| -- Tasty
    | -- New York
    | -- Beijing
    | -- Tokyo
    | -- ...
| -- New York       
```

When I click this tag, I basically want to collect all photos/documents related to NY. It should contain all things in `Travel/New York` and `Tasty/New York` and even some photos with just a single `New York` tag. But in your design, they are three different tags, even with the exact same name. It means that when we click the top level `New York` we will get nothing from `Travel/New York` and `Tasty/New York`.

That is why we need **atomicity**. If we have atomicity, all `New York`s are the same one, i.e., in human language `New York related`. When we click the top-level `New York`, we are collecting New York-related things. When we click the `Tasty - New York`, we are collecting New York-related and food-related things. In mathematics, it's getting the intersection between the `Tasty` set and the `New York` set.

Similarly, when we use the advanced search, we use a query like this: `'New York' in tags`, we should get all things related to `New York`.

---

If we are going to talk about the file/folder system, yes, your design is intuitive. But not for a tag system.

**Reply by woodwm** [reply]
No. I did not break the atomicity of tags. In my post, 

> `cv/segmentation` is a nested tag with two nested tag objects `cv` and `segmentation`.

Let me rename nested tag object as nested tag component. Every tag or nested tag are composed by nested tag component which includes the property of parient tag and child tag.

The only difference is that I think the tag component in different nested tags are different object with different UUID but the same tag name. `misc` does not require another tag `note` or `supplement` to differtiate with each other. As the UUID is hidden from the users, this means when the user add a tag `misc` to a paper, it is unclear which `misc` he refer to. Thus, the software need to provide suggestions (by searching in the tag component table) to allow the user to select `note\misc` or `supplement\misc`, and add the corresponding other part automatically. This process ensures the correctness of tag system.

When tags in different nested tag are the same, they can only be differentiat by its upper level tag. When we have 20 or 30 tags, it becomes a budden to maintain the tag system. You will always find that some papers missed some tags and then the nested tag is not correct. (This happens to me using Zotero. ) Eventually, every paper needs to have many tags to support the correct nested tag.

Both approach works and can show the same result with different approaches. No right or wrong, but different using patterns. I think taking tag component different in different tags is more easy to maitain the tag system. Potentially, I think it is also easy to search in the tree for a certain tag.

**Reply by charlieJ107** [reply]
Let me try to describe such a usage pain point to see if I understand why you want nested tags. When we collect a new paper, we may have some tags in mind, such as `misc`, `segmentation`, but forget to add other tags. For example, although we added the segmentation tag to a paper, because I did not Adding either of the two tags `cv` or `multi-modal` will cause me to not know whether this paper is a `segmentation` in a `cv` context or a `segmentation` in a `multi-modal` context when I search in the future. Based on such a usage pain point, you want to introduce nested tag, and when the user tries to add a tag named `segmentation`, force the user to choose one of `cv` or `multi-modal`, and ensure that these papers identify the `segmentation` tag. It must be represented by one of the two tags `cv` or `multi-modal`. Is this a mechanism you want?

**Reply by GeoffreyChen777** [reply]
Here we just talk about user aspect design. For implementation, please refer to https://github.com/Future-Scholars/paperlib/discussions/354#discussioncomment-8162974. We are designing a graph-based implementation, and we can use a graph to generate the final query for the database. It is the work for our teams because users may know nothing about coding.

For user aspect. What we should discuss are:
- what do we want to see in the sidebar tree?
- what do we want to see in the details panel?
- how to add new tags?

## What do we want to see in the sidebar tree?

I just have one question.
```
| -- Travel
    | -- New York
    | -- Chicago
| -- Tasty
    | -- New York
    | -- Beijing
    | -- Tokyo
    | -- ...
| -- New York 
```

When I click the top-level tag `New York`, do you want to get the results including things under `Travel/New York` and `Tasty/New York`?

If your answer is yes. Then you should realise that these three tags are identical. Because they represent the same thing: "New York related". They are at different positions of the tree. The position only indicates if there are any additional conditions. The postion will never change the meaning of this tag itself. Don't consider the code implementation, just from user experience. 
If your answer is no. Then what you want is a file/folder system, not a tag system. Please understand what is folder and what is tag.

## what do we want to see in the details panel?

In the details panel, which one do you want to see?
- `Travel / New York`. : a single nested tag string
- `Travel` `New York` : atomic tag string.

If you choose the first one, then you want a file/folder system. Because tags should be atomic, `Travel / New York` can be split into two tags, which breaks the atomicity.

## how to add new tags?

when we open the editing view. we have an input box. If we type `New York` and then directly press Enter, it means that we assign the top-level `New York`. if we type `Travel/New York` and press Enter, it means that we what to assign two atomic tags: `Travel` and `New York`, and specifically, it should be the `New York` under `Travel`. However this `New York` is the same as the top-level `New York`. The main difference is that they are two copies of the same object/concept at different positions in the tree. The position only indicates that how to filter the database. The position of a tag in a tree will never change its meaning. The `New York` tag at anywhere always represents "New York related". So we just keep one single object in the database.

---

I think what you want is a file/folder system, not a tag system. Anyway, you can use a folder system as a 'tag' system, it depends on you.

**Reply by GeoffreyChen777** [reply]
@woodwm Hi, as you mentioned the tag system in Obsidian, I tried it and understand how they design their tag system.

Basically, it's a folder system and has no atomicity. I.e., `New York` is different from `Travel/New York`.

In the current version, we have tags, folders, and smartfilters. In Paperlib's next step, we will still have a folder system, **but is nested**. Compared with the file system in an OS, here the folder system allows you to link a paper to multiple folders without copying the source file. That would give you an exactly similar experience with Obsidian's tag system.

For the next steps of tag system and smartfilter system, how to design it would not change your usage preference with the help of the nested folder system.

Thanks for all of your constructive comments and have a good weekend.

**Reply by woodwm** [reply]
Both designs should work. Users will choose what is best suitable for him/her. Please go ahead in your plan. Thanks.

**Comment by GeoffreyChen777** [comment]
Hi @woodwm @fkcptlst

I've implemented the nested folders and even smartfilters #360. The code has been merged into the `dev-3.0` branch and this feature will be released with the next 3.0 beta version.

<img width="1120" alt="image" src="https://github.com/Future-Scholars/paperlib/assets/14183213/824f10ba-6ca9-4727-bef8-f43830e07646">

Thanks again for your constructive comments. Now I'm closing this discussion.
